File: src/Yitter.IdGenerator/SnowWorkerM2.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Yitter.IdGenerator
{
    /// <summary>
    /// 常规雪花算法
    /// </summary>
    internal class SnowWorkerM2 : SnowWorkerM1
    {
        public SnowWorkerM2(IdGeneratorOptions options) : base(options)
        {

        }

        public override long NextId()
        {
            lock (_SyncLock)
            {
                long currentTimeTick = GetCurrentTimeTick();

                if (_LastTimeTick == currentTimeTick)
                {
                    if (_CurrentSeqNumber++ > MaxSeqNumber)
                    {
                        _CurrentSeqNumber = MinSeqNumber;
                        currentTimeTick = GetNextTimeTick();
                    }
                }
                else
                {
                    _CurrentSeqNumber = MinSeqNumber;
                }

                if (currentTimeTick < _LastTimeTick)
                {
                    throw new Exception(string.Format("Time error for {0} milliseconds", _LastTimeTick - currentTimeTick));
                }

                _LastTimeTick = currentTimeTick;
                var result = ((currentTimeTick << _TimestampShift) + ((long)WorkerId << SeqBitLength) + (uint)_CurrentSeqNumber);

                return result;
            }
        }

    }
}

File: C/source/idgen/SnowWorkerM2.c
/*
 * 版权属于：yitter(yitter@126.com)
 * 代码翻译：amuluowin
 * 代码修订：yitter
 * 开源地址：https://gitee.com/yitter/idgenerator
 */
#include <malloc.h>
#include <stdlib.h>
#include <pthread.h>
#include "SnowWorkerM2.h"


extern uint64_t WorkerM2NextId(SnowFlakeWorker *worker) {
    pthread_mutex_lock(&ThreadMutex);
    uint64_t currentTimeTick = GetCurrentTimeTick(worker);
    if (worker->_LastTimeTick == currentTimeTick) {
        worker->_CurrentSeqNumber = (++worker->_CurrentSeqNumber) & worker->MaxSeqNumber;
        if (worker->_CurrentSeqNumber == 0) {
            currentTimeTick = GetNextTimeTick(worker);
        }
    } else {
        worker->_CurrentSeqNumber = worker->MinSeqNumber;
    }
    worker->_LastTimeTick = currentTimeTick;
    uint64_t id = (uint64_t) ((currentTimeTick << worker->_TimestampShift) |
                              (worker->WorkerId << worker->SeqBitLength) |
                              worker->_CurrentSeqNumber);
    pthread_mutex_unlock(&ThreadMutex);
    return id;
}
